Replumbing: add fallback provider capability

levitte opened the pull request:
To ensure that old applications aren't left without any provider, and
at the same time not forcing any default provider on applications that
know how to deal with them, we device the concept of fallback
providers, which are automatically activated if no other provider is
already activated.

We fully expect to see something like this in libcrypto initialization:

    OSSL_PROVIDER *prov = NULL;

    if ((prov = ossl_provider_new(NULL, "default",
                                  default_provider_init)) != NULL
        && ossl_provider_set_fallback(prov))
        ...

UPDATE: prepopulating with whatever default fallbacks now happens
whenever a new provider store is created.  This automatically ties it to
the library context being used.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated
- [ ] tests are added or updated

levitte committed:
Replumbing: add fallback provider capability

To ensure that old applications aren't left without any provider, and
at the same time not forcing any default provider on applications that
know how to deal with them, we device the concept of fallback
providers, which are automatically activated if no other provider is
already activated.

mattcaswell commented on doc/internal/man3/ossl_provider_new.pod:
```diff
@@ -92,6 +94,9 @@ one was loaded, and the I<provider object> will be freed.
 provider module.
 
+ossl_provider_set_fallback() marks an available provider as fallback.
+If the given provider is already "activated", this has no effect.
```
Presumably it will remain a fallback in the event that it is subsequently unloaded, so it still has an effect in that case.

levitte replied:
True, but that also requires that *all* activated providers are deactivated at that point.

mattcaswell replied:
Right...its a minor point. Still approved.

levitte commented:
Test added.  Let's see what the CIs think of it

mattcaswell commented:
Travis failure looks relevant:

````
../../_srcdist/test/recipes/02-test_internal_provider.t ........ 

1..1

    # Subtest: ../../test/provider_internal_test

    1..1

    ok 1 - test_builtin_provider

provider_internal_test(34828,0x7fffab0ce340) malloc: *** error for object 0x7fb6b24101d0: incorrect checksum for freed object - object was probably modified after being freed.

*** set a breakpoint in malloc_error_break to debug

../../util/shlib_wrap.sh ../../test/provider_internal_test => 134

not ok 1 - running provider_internal_test
````

mattcaswell commented:
Aside from the above issue this looks ok.

mattcaswell commented:
> We fully expect to see something like this in libcrypto initialization:

Just wanted to point out that this has to happen on a per OPENSSL_CTX basis

levitte commented:
> Just wanted to point out that this has to happen on a per OPENSSL_CTX basis

True...  and that's going to be interesting, 'cause I assume some of the global init stuff might actually move to library context init.   But either way, `NULL` in the `ossl_provider_new` call is that argument, where `NULL` is the default library context, and could as well be an actual application allocated library context.

levitte commented:
Re the Travis failure, I did have some vague notion that this might happen.  I believe my test program is at fault...

levitte commented:
I've remove the changes that I did in `test/provider_internal_test.c` because the fallback mechanism isn't testable with just the internal API.  We cannot really test it before we have some minimal public API that uses providers in place, i.e. when we start having a default provider.

levitte committed:
Replumbing: Add a mechanism to pre-populate the provider store

OpenSSL will come with a set of well known providers, some of which
need to be accessible from the start.  These are typically built in
providers, or providers that will work as fallbacks.

We do this when creating a new provider store, which means that this
will happen in every library context, regardless of if it's the global
default one, or an explicitely created one.

We keep the data about the known providers we want to make accessible
this way in crypto/provider_predefined.h, which may become generated.
For now, though, we make it simple and edited manually.

levitte commented:
@mattcaswell, you might be interested in the last development...  I just added a mechanism to add predefined providers, which is especially interesting for fallback providers.

(yes, I know that we currently plan on just *one* fallback provider, but who knows what we decide in the future?  We might want to break down the default by algorithm...)

paulidale commented on crypto/provider_predefined.h:
```diff
@@ -0,0 +1,21 @@
+    const char *name;
+    OSSL_provider_init_fn *init;
+    unsigned int is_fallback:1;
+} const predefined_providers[] = {
```
Is it necessary to populate this array in an _header_ file?

How about having this in the header:
```
typedef struct {
...
} PREDEFINED_PROVIDER;

extern const PREDEFINED_PROVIDER predefined_providers[];
extern const size_t num_predefined_providers;
```

And this in a corresponding .c file:
```
const PREDEFINED_PROVIDER predefined_providers[] = {
    { "default", ossl_default_provider_init, 1 },
};

const size_t num_predefined_providers = OSSL_NELEM(predefined_providers);
```

levitte replied:
It's practical for generation purposes, and it's not like any other C unit should be interested in this particular file...

But hey, I could just put it all in `provider_core.c` if you prefer

paulidale replied:
I guess it doesn't matter much which way.

I cringe a bit at code/data in headers, unless they're machine generated.

levitte replied:
> I cringe a bit at code/data in headers, *unless they're machine generated.*

And that's precisely it. We might not generate it now, but this prepares us for a future where we do.

paulidale replied:
Okay, all fine if that's intended at some point.

richsalz replied:
I agree that it jars putting data in a header file and requires nobody else to ever include it. I think it would avoid questions if you moved it.

levitte replied:
It's not like this is anything new...  see `crypto/asn1/charmap.h`, `crypto/asn1/asn1_item_list.h`, `crypto/asn1/standard_methods.h`, `crypto/asn1/tbl_standard.h`, `crypto/x509v3/standard_exts.h`

But yeahok, it can be made into a .c file

paulidale commented on doc/internal/man3/ossl_provider_new.pod:
```diff
@@ -84,14 +86,20 @@ To indicate a built-in provider, the C<init_function> argument must
-if it drops down to zero, the associated module will be unloaded if
-one was loaded, and the I<provider object> will be freed.
+counter; if it drops below 2, the I<provider object> is assumed to
+have fallen out of use and will be inactivated (it's teardown function
```
_its_ not _it's_.  Deactivated rather than inactivated.

levitte replied:
Will fix

levitte replied:
(found another one too)

levitte replied:
Found another one too

mattcaswell commented on crypto/provider_core.c:
```diff
@@ -109,23 +115,34 @@ static struct provider_store_st *get_provider_store(OPENSSL_CTX *libctx)
+                                predefined_providers[i].init);
+
+            if (prov == NULL
+                || sk_OSSL_PROVIDER_push(store->providers, prov) == 0) {
```
Doesn't this need to be locked?

levitte replied:
Actually, yes.  Thanks

mattcaswell commented on crypto/provider_core.c:
```diff
@@ -336,23 +375,69 @@ int ossl_provider_forall_loaded(OPENSSL_CTX *ctx,
+             * if use_fallbacks is 1, it means there were no activated
+             * providers at all, and we need to auto-activate any fallback
+             * and try once more.
+             */
```
I'm wondering if this is the right logic. It doesn't seem possible with how things are right now to be able to say "never use the default provider under any circumstances". I'm thinking that (perhaps) we should say that as soon as one provider has been activated we should never automatically activate the default provider even if the first provider is later de-activated, i.e. we only fallback if we have *never* had an active provider.

levitte replied:
Fair point.  That's easily solved with a flag in the store itself (i.e. in `struct provider_store_st`.

levitte committed:
fixup! Replumbing: add fallback provider capability

levitte committed:
fixup! Replumbing: Add a mechanism to pre-populate the provider store

levitte committed:
fixup! Replumbing: add fallback provider capability

levitte committed:
fixup! Replumbing: add fallback provider capability

levitte committed:
fixup! Replumbing: add fallback provider capability

mattcaswell commented:
Looks like there is a "make update" discrepancy (hence the travis red cross)

levitte commented:
> Looks like there is a "make update" discrepancy (hence the travis red cross)

Ah, right.  Fixing it now

levitte committed:
fixup! Replumbing: Add a mechanism to pre-populate the provider store

levitte committed:
fixup! Replumbing: add fallback provider capability

levitte committed:
fixup! Replumbing: Add a mechanism to pre-populate the provider store

levitte committed:
fixup! Replumbing: Add a mechanism to pre-populate the provider store

mattcaswell commented on crypto/provider_core.c:
```diff
@@ -107,25 +116,41 @@ static struct provider_store_st *get_provider_store(OPENSSL_CTX *libctx)
+                    ossl_provider_set_fallback(prov);
+            }
+        }
+        CRYPTO_THREAD_unlock(store->lock);
```
So every time we do a fetch or touch the store we will obtain a write lock for a short time? I think this could be a big source of lock contention. But actually this code only ever needs to be done once - at the time the store is created. Wouldn't it be better placed in `provider_store_new` to avoid the lock?

levitte replied:
True.  Ok, hold on, will experiment with that

levitte replied:
Done

mattcaswell commented on crypto/provider_core.c:
```diff
@@ -329,6 +382,46 @@ int ossl_provider_activate(OSSL_PROVIDER *prov)
 
+int ossl_provider_activate(OSSL_PROVIDER *prov)
+{
+    if (provider_activate(prov)) {
```
The comment above says `provider_activate` is the internal lockless version - but here we're just calling it anyway without obtaining a lock??

levitte replied:
The activation in itself doesn't need protection, it's already assumed that the reference count makes it legitimate to poke the activated flag.  The lock happens below, to set the store flag.

mattcaswell replied:
In which case the comment above `provider_activate` is probably misleading and could do with rewording

levitte replied:
Okie

levitte replied:
Done

levitte committed:
fixup! Replumbing: Add a mechanism to pre-populate the provider store

levitte committed:
fixup! Replumbing: Add a mechanism to pre-populate the provider store

levitte committed:
fixup! Replumbing: add fallback provider capability

mattcaswell commented on crypto/provider_core.c:
```diff
@@ -339,20 +428,54 @@ int ossl_provider_forall_loaded(OPENSSL_CTX *ctx,
+             */
+            ret = provider_forall_loaded(store, NULL, cb, cbdata);
+
+            /* TODO: should we set store->use_fallbacks = 0 here? */
```
IMO - yes we should. IIUC, it should normally be 0 here anyway (because we will normally always have at least one fallback). The only way we should get here is if we have already attempted to active fallbacks, but we still didn't end up activating any providers. In that case it makes sense to set use_fallbacks to 0 anyway because doing it again seems unlikely to succeed a second time.

levitte replied:
I'm imagining the following scenario:

1. an algorithm is fetched, and since no provider is activated, "default" gets activated as a fallback.
2. another fallback provider is added (let's call it "foo").
3. an algorithm is fetched, isn't found among current providers (i.e. "default", as that's the only one activated so far), but would be found if "foo" was activated.

Of course, that scenario isn't possible if we decide that *all* fallbacks must be in place before algorithm fetches are done.

mattcaswell replied:
Surely in that scenario `use_fallbacks` would get set to 0 as soon as `default` got activated in step 1. Therefore we never attempt to activate fallbacks again after that point so we would never attempt to active "foo". IMO fallbacks are mostly only interesting in scenarios where applications are not provider aware. As soon as an application start dynamically adding and removing providers then we assume it knows what its doing and fallbacks shouldn't be necessary.

levitte replied:
Note that fallbacks can only be set internally...  I guess I'm just try to think forward of possible corner cases...  But you know what, I can always set use_fallbacks to zero here with an added comment that we might want to reconsider it.

levitte commented on crypto/provider_core.c:
```diff
@@ -75,13 +83,36 @@ static void provider_store_free(void *vstore)
+            || sk_OSSL_PROVIDER_push(store->providers, prov) == 0) {
+            ossl_provider_free(prov);
+            CRYPTOerr(CRYPTO_F_PROVIDER_STORE_NEW, ERR_R_INTERNAL_ERROR);
+            return NULL;
```
Yes!  Good catch, thank you

levitte commented on crypto/provider_core.c:
```diff
@@ -164,18 +214,8 @@ OSSL_PROVIDER *ossl_provider_new(OPENSSL_CTX *libctx, const char *name,
-        ossl_provider_free(prov);
-        CRYPTOerr(CRYPTO_F_OSSL_PROVIDER_NEW, ERR_R_MALLOC_FAILURE);
+    if ((prov = provider_new(name, init_function)) == NULL)
         return NULL;
```
Sure

levitte commented on crypto/provider_core.c:
```diff
@@ -190,6 +230,8 @@ OSSL_PROVIDER *ossl_provider_new(OPENSSL_CTX *libctx, const char *name,
     if (prov == NULL)
         CRYPTOerr(CRYPTO_F_OSSL_PROVIDER_NEW, ERR_R_MALLOC_FAILURE);
+    else
+        prov->store = store;
```
I was actually unsure about that...  will do.

levitte committed:
fixup! Replumbing: Add a mechanism to pre-populate the provider store

levitte committed:
fixup! Replumbing: add fallback provider capability

levitte committed:
fixup! Replumbing: add fallback provider capability

levitte commented:
Merged.  Thanks for the review!

c41f3ae0d92a87b903a9ed585622adae06791676 Replumbing: Add a mechanism to pre-populate the provider store
e55008a9f2ae299374dcf868b660389e84dd2e0b Replumbing: add fallback provider capability
